usdt-impl/src/common.rs

```diff
@@ -37,23 +56,37 @@ pub fn generate_type_check(types: &[DataType]) -> TokenStream {
 
     let preamble = unpack_argument_lambda(&types);
 
-    let generics = match (has_strings, has_serializables) {
-        (false, false) => quote! {},
-        (true, false) => quote! { <S: AsRef<str>> },
-        (false, true) => quote! { <T: ::serde::Serialize> },
-        (true, true) => quote! { <S: AsRef<str>, T: ::serde::Serialize> },
+    // NOTE: We currently are a bit loose with string types. In particular, a probe function
+    // defined with an argument of type `String` or `&str` map to the same `DataType` variant. The
+    // argument closure can actually be called with a _different_ type, as long as that type
+    // implements `AsRef<str>`. For example, a probe defined like:
+    //
+    // ```rust
+    // #[usdt::provider]
+    // mod provider {
+    //      fn probe(_: String) {}
+    // }
+    // ```
+    //
+    // Can be called like:
+    //
+    // ```rust
+    // provider_probe!(|| "This is a `&'static str`, not a `String`");
+    // ```
+    let generics = if has_strings {
+        quote! { <S: AsRef<str>> }
+    } else {
+        quote! {}
     };
 
-    // NOTE: This block defines an internal empty function and then a lambda which
-    // calls it. This is all strictly for type-checking, and is optimized out. It is
-    // defined in a scope to avoid multiple-definition errors in the scope of the macro
-    // expansion site.
+    let type_check_function =
+        format_ident!("__usdt_private_{}_{}_type_check", provider_name, probe_name);
     quote! {
```

Comment:
again, I think if we put this all under `quote_spanned!` it might have a slightly better ui for the error message.